# Your CNC Machine Knows More Than GPT-5.4
## The Industry's Blind Spot in AGI Benchmarks

**Track**: Attention | **Team**: U2DIA | **Competition**: Kaggle Measuring AGI

---

After 23 years of operating CNC machines, building automated production lines, and programming 5-axis mills and industrial robots by hand, I started building AI systems to automate what I used to do manually.

That is when I discovered a paradox: **the same AI that can generate a working CNC program cannot spot a single wrong tool offset in the program it just wrote.**

A skilled machinist catches it in seconds. The best frontier model misses it 40% of the time.


In [1]:
import pandas as pd
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import numpy as np


## Model Performance — 7 Models × 5 Domains × 53 Tasks


In [2]:
data = {
    'Model': ['Claude Sonnet 4.6', 'GPT-5.4', 'Llama 3.3 70B', 'Gemini 3 Flash',
              'GPT-4.1-mini', 'GPT-4o', 'Llama 3.2 3B'],
    'Overall': [79.1, 77.1, 65.8, 65.0, 60.8, 58.5, 17.2],
    'G-code': [88.9, 85.4, 77.4, 71.9, 52.6, 65.8, 26.1],
    'Cutting Force': [90.0, 92.5, 67.5, 75.0, 85.0, 77.5, 0.0],
    'Physics': [76.0, 76.0, 52.0, 68.0, 72.0, 68.0, 16.0],
    'Solvers': [100.0, 100.0, 90.0, 82.5, 95.0, 92.5, 0.0],
    'Human Corr.': [16.0, 14.0, 6.0, 14.0, 7.0, 6.0, 10.0],
}
df = pd.DataFrame(data).set_index('Model')
df.style.background_gradient(cmap='RdYlGn', vmin=0, vmax=100).format('{:.1f}%')


,Overall,G-code,Cutting Force,Physics,Solvers,Human Corr.
Model,,,,,,
Claude Sonnet 4.6,79.1%,88.9%,90.0%,76.0%,100.0%,16.0%
GPT-5.4,77.1%,85.4%,92.5%,76.0%,100.0%,14.0%
Llama 3.3 70B,65.8%,77.4%,67.5%,52.0%,90.0%,6.0%
Gemini 3 Flash,65.0%,71.9%,75.0%,68.0%,82.5%,14.0%
GPT-4.1-mini,60.8%,52.6%,85.0%,72.0%,95.0%,7.0%
GPT-4o,58.5%,65.8%,77.5%,68.0%,92.5%,6.0%
Llama 3.2 3B,17.2%,26.1%,0.0%,16.0%,0.0%,10.0%


## Heatmap — Where Do Models Fail?


In [3]:
fig, ax = plt.subplots(figsize=(12, 5))
im = ax.imshow(df.values, cmap='RdYlGn', aspect='auto', vmin=0, vmax=100)
ax.set_xticks(range(len(df.columns)))
ax.set_xticklabels(df.columns, rotation=25, ha='right', fontsize=11)
ax.set_yticks(range(len(df.index)))
ax.set_yticklabels(df.index, fontsize=11)
for i in range(len(df.index)):
    for j in range(len(df.columns)):
        v = df.values[i, j]
        ax.text(j, i, f'{v:.0f}%', ha='center', va='center',
                fontsize=10, fontweight='bold', color='black' if v > 50 else 'white')
plt.colorbar(im, ax=ax, label='Pass Rate (%)', shrink=0.8)
ax.set_title('Industrial Attention Benchmark — 53 Tasks x 7 Models', fontsize=14, pad=15)
plt.tight_layout()
plt.show()


## Key Insights

### 1. Human Corrections Are the Great Equalizer
Even Claude Sonnet 4.6 scores only **16%** on tasks derived from real AI→human code corrections. These are errors that an AI actually made in production, and its successors still can't spot them.

### 2. Industrial Domains Expose Genuine Gaps
The 62.8pp spread on G-code tasks exists because CNC G-code simply isn't in the training distribution. Models that "understand code" by extrapolating from Python/JS systematically misread industrial syntax.

### 3. Counting Is Still Broken
Even GPT-5.4 confabulates tool counts in 80+ line G-code programs — listing matches correctly but miscounting by 1-2.

### 4. The Paradox
**AI can generate G-code. AI cannot review G-code.** A skilled machinist with 23 years of experience catches a wrong tool offset in seconds. The best frontier model misses it 21% of the time.


## Theoretical Foundation

Our benchmark maps to DeepMind's three Attention dimensions (Burnell et al., 2026):

| Dimension | Template | Cognitive Analog |
|-----------|----------|------------------|
| **Capacity** | Tool counting (att-02) | Kahneman (1973) — finite shared resource |
| **Selective Control** | Error localization (att-01/03/05/06) | Eriksen flanker (1974) |
| **Stimulus-driven** | Outlier detection (att-04) | Treisman attenuation (1964) |

---

## About

Built from 28,214 real Fanuc CNC programs, 500,000+ lines of AI-authored code, and 23 years of manufacturing experience.

If you find this useful, **please upvote!**
